sdk/communication/Azure.Communication.Common/src/EntraTokenCredential.cs

```diff
@@ -0,0 +1,145 @@
+    internal sealed class EntraTokenCredential : ICommunicationTokenCredential
+    {
+        private HttpPipeline _pipeline;
+        private string _resourceEndpoint;
+        private readonly ThreadSafeRefreshableAccessTokenCache _accessTokenCache;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="EntraTokenCredential"/>.
+        /// </summary>
+        /// <param name="options">The options for how the token will be fetched</param>
+        /// <param name="pipelineTransport">Only for testing.</param>
+        public EntraTokenCredential(EntraCommunicationTokenCredentialOptions options, HttpPipelineTransport pipelineTransport = null)
+        {
+            this._resourceEndpoint = options.ResourceEndpoint;
+            _pipeline = CreatePipelineFromOptions(options, pipelineTransport);
+            _accessTokenCache = new ThreadSafeRefreshableAccessTokenCache(
+                    ExchangeEntraToken,
+                    ExchangeEntraTokenAsync,
+                    false, null, null);
+            _accessTokenCache.GetValueAsync(default);
+        }
+
+        private HttpPipeline CreatePipelineFromOptions(EntraCommunicationTokenCredentialOptions options, HttpPipelineTransport pipelineTransport)
+        {
+            var authenticationPolicy = new BearerTokenAuthenticationPolicy(options.TokenCredential, options.Scopes);
+            var entraTokenGuardPolicy = new EntraTokenGuardPolicy();
+            var clientOptions = ClientOptions.Default;
+            if (pipelineTransport != null)
+            {
+                clientOptions.Transport = pipelineTransport;
+            }
+            return HttpPipelineBuilder.Build(clientOptions, authenticationPolicy, entraTokenGuardPolicy);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            _pipeline = default;
+            _accessTokenCache.Dispose();
+        }
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns> Contains the access token.</returns>
+        public AccessToken GetToken(CancellationToken cancellationToken = default)
+            => _accessTokenCache.GetValue(cancellationToken, () => true);
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns>
+        /// A task that represents the asynchronous get token operation. The value of its <see cref="ValueTask{AccessToken}.Result"/> property contains the access token.
+        /// </returns>
+        public ValueTask<AccessToken> GetTokenAsync(CancellationToken cancellationToken = default)
+            => _accessTokenCache.GetValueAsync(cancellationToken, () => true);
+
+        private AccessToken ExchangeEntraToken(CancellationToken cancellationToken)
```

Comment:
I don't see anything calling this method?

Reply:
32th line, it's provided as a param to the ThreadSafeRefreshableAccessTokenCache init